### PR TITLE
[WIP] Recycle unused FiberCache directly in OapCache

### LINKED ISSUE #827: Multi-thread CacheGuardian
CacheGuardian, performing off-heap memory recycling, currently uses 1 thread.

Will the not used Fiber be recycled more promptyly using multi-thread?

### PR DESCRIPTION
**Pull request by gczsjdy**
## What changes were proposed in this pull request?

To resolve #827  This removed CacheGuardian and recycle the memory more directly. 

## How was this patch tested?

Existing tests

### CODE REVIEW

**Comment by gczsjdy** [comment]
@LuciferYang @linhong-intel Is this the way you meant? But I'm afraid too many `onRemoval` calls + a lot of `Fiber` in use will cause many threads' frequent context switch...

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala [hunk 1]
```diff
@@ -110,26 +106,33 @@ class SimpleOapCache extends OapCache with Logging {
...
   // Total cached size for debug purpose, not include pending fiber
   private val _cacheSize: AtomicLong = new AtomicLong(0)
 
+  private val fiberCacheReleasePool = ExecutionContext.fromExecutorService(
+    ThreadUtils.newDaemonCachedThreadPool(
+      "OAP-release-FiberCache", Runtime.getRuntime.availableProcessors()))
+
+  private val releaseMemoryThreadSleepingInterval = 200
+
   private val removalListener = new RemovalListener[FiberId, FiberCache] {
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala]: This should be still synchronized, the official usage is use `RemovalListeners.asynchronous(removalListener ,fiberCacheReleasePool)`, then register the `return value` to CacheBuilder.....

Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala]: Thanks ! : )

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala [hunk 2]
```diff
@@ -76,20 +75,31 @@ trait OapCache {
     cache
   }
 
+  protected val fiberCacheReleasePool = ExecutionContext.fromExecutorService(
+    ThreadUtils.newDaemonCachedThreadPool(
+      "OAP-release-FiberCache", Runtime.getRuntime.availableProcessors()))
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala]: We create a cached thread pool whose max number of threads is `Runtime.getRuntime.availableProcessors())` and keepAlive 60s，should we let the `maxThreadNumber` configable?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala [hunk 3]
```diff
@@ -197,7 +212,7 @@ class GuavaOapCache(cacheMemory: Long, cacheGuardianMemory: Long) extends OapCac
     CacheStats(
       dataFiberCount.get(), dataFiberSize.get(),
       indexFiberCount.get(), indexFiberSize.get(),
-      pendingFiberCount, cacheGuardian.pendingFiberSize,
+      0, 0,
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala]: Don't we need to record pendingFiberCount and pendingFiberSize ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala [hunk 4]
```diff
@@ -19,14 +19,14 @@ package org.apache.spark.sql.execution.datasources.oap.filecache
...
-
 import com.google.common.cache._
-
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql.execution.datasources.OapException
 import org.apache.spark.sql.oap.OapRuntime
-import org.apache.spark.util.Utils
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+import scala.collection.JavaConverters._
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala]: This 2 line should be moved to line 22...

**Comment by carsonwang** [comment]
I think this does not work very well. When you reach the max thread size, new tasks will be put to the queue and do not have the chance to release the fiber until one thread is free. Unfortunately the threads may all waiting for the fibers being released and do not finish quickly. This just like you are waiting for some fibers being release while the released fibers can't be recycled.

**Comment by LuciferYang** [comment]
So do we need to set maxThreadNumber to Runtime.getRuntime.availableProcessors()? use default value?

**Pull request closed without merging**